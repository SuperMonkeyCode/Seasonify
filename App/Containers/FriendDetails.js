import React from 'react';
import { connect } from 'react-redux';
import { Keyboard } from 'react-native';
import {
  Container,
  Content,
  Form,
  Label,
  Item,
  Input,
  Icon,
} from 'native-base';

import { friendActions } from '../Redux/FriendRedux';
import HeaderModular from '../Components/HeaderModular';
import { successMessage, warningMessage } from '../Config/DefaultMessages';

const mapStateToProps = null; // state => ({});

const mapActionsToProps = {
  removeFriend: friendActions.removeFriend,
  updateFriend: friendActions.updateFriend,
};

export class FriendDetails extends React.Component {
  static navigationOptions = ({ navigation }) => {
    const { params } = navigation.state;
    return {
      header: (
        <HeaderModular
          title={params.friend.name}
          cancelButton={{ icon: 'arrow-back', action: navigation.goBack }}
          actionButtons={[
            {
              visibleIf: !params.isEditing,
              icon: 'create',
              action: params.handleEdit,
            },
            {
              visibleIf: params.isEditing,
              hideByDefault: true,
              icon: 'checkmark',
              action: params.handleDone,
            },
            { icon: 'trash', action: params.handleDelete },
          ]}
        />
      ),
    };
  };

  componentWillMount() {
    // local model
    this.setState({ friend: this.props.navigation.state.params.friend });

    // actions and params for the navigation state
    this.props.navigation.setParams({
      isEditing: false,
      handleEdit: this.handleEdit,
      handleDone: this.handleDone,
      handleDelete: this.handleDelete,
    });
  }

  onChangeName = name => {
    this.syncModel({ name });
  };

  syncModel = ({ name }, callback = () => {}) => {
    this.setState(
      {
        friend: { ...this.state.friend, name: name.trim() },
      },
      callback,
    );
    this.props.navigation.setParams({
      friend: { ...this.props.navigation.state.params.friend, name },
    });
  };

  handleEdit = () => {
    this.props.navigation.setParams({
      isEditing: true,
    });
  };

  handleDone = () => {
    Keyboard.dismiss();
    this.props.updateFriend(this.state.friend);
    this.props.navigation.navigate('FriendList', {
      message: {
        ...successMessage,
        text: `${this.state.friend.name} has been edited`,
      },
    });
  };

  handleDelete = () => {
    const { friend } = this.props.navigation.state.params;
    this.props.removeFriend(friend.id);
    this.props.navigation.navigate('FriendList', {
      message: {
        ...warningMessage,
        text: `${friend.name} has been deleted`,
      },
    });
  };

  render() {
    const { params } = this.props.navigation.state;

    return (
      <Container>
        <Content>
          <Form>
            <Item fixedLabel>
              <Label>Name</Label>
              {params.isEditing && <Icon name="create" />}
              <Input
                disabled={!params.isEditing}
                value={params.friend.name}
                onChangeText={this.onChangeName}
                autoCapitalize="words"
              />
            </Item>
          </Form>
        </Content>
      </Container>
    );
  }
}

export default connect(mapStateToProps, mapActionsToProps)(FriendDetails);
