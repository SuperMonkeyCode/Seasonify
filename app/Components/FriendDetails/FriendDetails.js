/* @flow */
import {
  Body,
  Button,
  Container,
  Content,
  Form,
  Icon,
  Input,
  Item,
  Label,
  Left,
  ListItem,
  Picker,
} from 'native-base';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';

import type { Friend } from 'Types';
import cssColors from 'Themes/cssColors';

import styles from './FriendDetails.style';

type Props = {
  /* parent */
  navigation: Object,
  /* connect */
  friend: Friend,
  isEditing: boolean,
  editedObject: Object,
  editUpdate: Function,
  friendUpdate: Function,
  /* HOC */
  handleChangeName: Function,
  handleChangeColor: Function,
};

const enhance = compose(
  pure,
  withHandlers({
    handleChangeName: ({ navigation, editUpdate }: Props) => (name: string) => {
      editUpdate({ id: navigation.state.params.friendId, name });
    },
    handleChangeColor: ({ navigation, friendUpdate }: Props) => (color: string) => {
      friendUpdate({ id: navigation.state.params.friendId, color });
    },
  }),
);

function FriendDetails({
  friend,
  isEditing,
  editedObject,
  handleChangeName,
  handleChangeColor,
}: Props) {
  if (friend) {
    return (
      <Container>
        <Content>
          <Form>
            <Item fixedLabel>
              <Label>Name</Label>
              {isEditing && <Icon name="create" />}
              <Input
                disabled={!isEditing}
                value={'name' in editedObject ? editedObject.name : friend.name}
                onChangeText={handleChangeName}
                autoCapitalize="words"
              />
            </Item>
          </Form>
          <ListItem icon>
            <Left>
              <Button
                disabled
                style={{
                  ...styles.colorPatch,
                  backgroundColor: friend.color,
                }}
              />
            </Left>
            <Body>
              <Picker
                note
                iosHeader="Select a color"
                selectedValue={friend.color}
                onValueChange={handleChangeColor}
                mode="dropdown"
              >
                {Object.keys(cssColors).map((color: string) =>
                  <Picker.Item key={color} label={color} value={color} />,
                )}
              </Picker>
            </Body>
          </ListItem>
        </Content>
      </Container>
    );
  }
  return <Container />;
}

export default enhance(FriendDetails);
