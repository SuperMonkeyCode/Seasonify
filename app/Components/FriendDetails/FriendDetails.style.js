export default {
  colorPatch: {
    flexBasis: '29%',
  },
};
