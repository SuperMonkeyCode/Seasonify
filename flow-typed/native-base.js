/* @flow */
declare module 'native-base' {
  declare var exports: any;
}
