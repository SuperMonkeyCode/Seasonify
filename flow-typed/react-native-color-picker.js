/* @flow */
declare module 'react-native-color-picker' {
  declare var exports: any;
}
