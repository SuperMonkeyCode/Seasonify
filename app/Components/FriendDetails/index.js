/* @flow */
import { connect } from 'react-redux';
import React from 'react';

import { editActions, editSelectors } from 'State/editState';
import { friendSelectors } from 'State/friendState';
import FriendDetailsHeader from 'Components/FriendDetailsHeader';

import FriendDetails from './FriendDetails';

FriendDetails.navigationOptions = ({ navigation }) => ({
  header: (
    <FriendDetailsHeader
      friendId={navigation.state.params.friendId}
      navigation={navigation}
    />
  ),
});

export default connect(
  (state, ownProps) => ({
    friend: friendSelectors.getFriend(state, ownProps.navigation.state.params.friendId),
    isEditing: editSelectors.isEditing(state),
    editedObject: editSelectors.editedObject(state),
  }),
  {
    editUpdate: editActions.editUpdate,
  },
)(FriendDetails);
