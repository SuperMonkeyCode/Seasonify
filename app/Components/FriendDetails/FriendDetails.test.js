import { Button, Container, Input } from 'native-base';
import { ColorPicker } from 'react-native-color-picker';
import React from 'react';

import FriendDetails from './FriendDetails';

function setup(specificProps = {}) {
  const props = {
    navigation: {
      navigate: jest.fn(),
      state: { params: { friendId: 'abc123' } },
    },
    friend: {
      id: 'abc123',
      name: 'A friend',
      color: 'pink',
    },
    isEditing: false,
    editedObject: {},
    editUpdate: jest.fn(),
    friendUpdate: jest.fn(),
    ...specificProps,
  };
  const parent = shallow(<FriendDetails {...props} />);
  const level = [];
  level[0] = parent;
  level[1] = parent.dive();
  level[2] = level[1].dive();
  const component = level[2];
  return {
    component,
    parent,
    level,
    props,
    colorButton: component.find(Button).first(),
    inputs: component.find(Input),
  };
}

describe('Rendering when default state', () => {
  const { component } = setup();
  it('should match', () => {
    expect(component).toMatchSnapshot();
  });
});

describe('Rendering when no friend', () => {
  const { component } = setup({ friend: undefined });
  it('should match', () => {
    expect(component).toMatchSnapshot();
  });
});

describe('Rendering when editing', () => {
  const { component } = setup({
    isEditing: true,
    editedObject: { name: 'A friend mod' },
  });
  it('should match', () => {
    expect(component).toMatchSnapshot();
  });
});

describe('When picker visibility is true', () => {
  const { component } = setup();
  component.setProps({ isPickerVisible: true });

  it('displays the color picker', () => {
    expect(component.find(ColorPicker)).toHaveLength(1);
  });
});

/* ========= Events & Functions ========= */

describe('Events & Functions', () => {
  const { parent, props, colorButton, inputs } = setup();

  describe('When clicking the color picker', () => {
    colorButton.props().onPress();

    it('updates the picker visibility state', () => {
      expect(parent.state().isPickerVisible).toEqual(true);
    });
  });

  describe('when calling onChangeName', () => {
    inputs.first().props().onChangeText('New name');

    it('calls editUpdate', () => {
      expect(props.editUpdate).toBeCalledWith({
        id: props.friend.id,
        name: 'New name',
      });
    });
  });
});
