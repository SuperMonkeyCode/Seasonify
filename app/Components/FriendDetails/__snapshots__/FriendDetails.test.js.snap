// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering when default state should match 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Styled(Form)>
      <Styled(Item)
        fixedLabel={true}
      >
        <Styled(Label)>
          Name
        </Styled(Label)>
        <Styled(Input)
          autoCapitalize="words"
          disabled={true}
          onChangeText={[Function]}
          value="A friend"
        />
      </Styled(Item)>
    </Styled(Form)>
    <Styled(ListItem)
      icon={true}
    >
      <Styled(Left)>
        <Styled(Button)
          disabled={true}
          style={
            Object {
              "backgroundColor": "pink",
              "flexBasis": "29%",
            }
          }
        />
      </Styled(Left)>
      <Styled(Body)>
        <Styled(PickerNB)
          iosHeader="Select a color"
          mode="dropdown"
          note={true}
          onValueChange={[Function]}
          selectedValue="pink"
        >
          <Item
            label="pink"
            value="pink"
          />
          <Item
            label="blue"
            value="blue"
          />
        </Styled(PickerNB)>
      </Styled(Body)>
    </Styled(ListItem)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`Rendering when editing should match 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Styled(Form)>
      <Styled(Item)
        fixedLabel={true}
      >
        <Styled(Label)>
          Name
        </Styled(Label)>
        <Styled(Icon)
          name="create"
        />
        <Styled(Input)
          autoCapitalize="words"
          disabled={false}
          onChangeText={[Function]}
          value="A friend mod"
        />
      </Styled(Item)>
    </Styled(Form)>
    <Styled(ListItem)
      icon={true}
    >
      <Styled(Left)>
        <Styled(Button)
          disabled={true}
          style={
            Object {
              "backgroundColor": "pink",
              "flexBasis": "29%",
            }
          }
        />
      </Styled(Left)>
      <Styled(Body)>
        <Styled(PickerNB)
          iosHeader="Select a color"
          mode="dropdown"
          note={true}
          onValueChange={[Function]}
          selectedValue="pink"
        >
          <Item
            label="pink"
            value="pink"
          />
          <Item
            label="blue"
            value="blue"
          />
        </Styled(PickerNB)>
      </Styled(Body)>
    </Styled(ListItem)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`Rendering when no friend should match 1`] = `<Styled(Container) />`;
